src/cpp/session/SessionSourceDatabase.cpp

```diff
@@ -1033,6 +1034,11 @@ void onRemoveAll()
 
 SEXP rs_getDocumentProperties(SEXP pathSEXP, SEXP includeContentsSEXP)
 {
+   if (!r::exec::isMainThread())
```

Comment:
Not necessarily for this PR, but I wonder if we should instead make an appropriate change to our `RS_REGISTER_CALL_METHOD()` macro here:

https://github.com/rstudio/rstudio/blob/2171c9fa3bd775a8ddcdc5d35085df1717c88829/src/cpp/r/include/r/RRoutines.hpp#L75-L87

That way, we could have a way of detecting when any registered method is called from a separate thread (which should never happen), as opposed to just the ones we remember to sprinkle this on.

Reply:
This could be useful but I'm going to hold off on this change for now because it's not obvious and will make tricky code even trickier :) I think that we mainly need these guards to prevent RStudio calls into the R runtime off the main thread. As I understand it, by the time these calls are made by the R runtime, we should have already caught the problem.  I do have checks the callbacks: polledEventHandler and rReadConsole but they turned out not to be that helpful since the stack was corrupted already by the time we got there on the wrong thread.

Reply:
Sounds good!